leetcode/src/23.c

```diff
@@ -0,0 +1,69 @@
+/**
+ * Definition for singly-linked list.
+ * struct ListNode {
+ *     int val;
+ *     struct ListNode *next;
+ * };
+ * 
+ * Algorithm merges pairwise lists by divide and conquer.
+ * Space complexity O(n * k), time complexity O(k * n * log(k)), where n - average size of list and k - number of lists.
+ */
+struct ListNode *merge_two(struct ListNode *ln1, struct ListNode *ln2) {
+    struct ListNode *head, *node;
+    if (ln1 == NULL) {
+        return ln2;
+    } else if (ln2 == NULL) {
+        return ln1;
+    }
+    if (ln1->val < ln2->val) {
+        head = ln1;
+        ln1 = ln1->next;
+    } else {
+        head = ln2;
+        ln2 = ln2->next;
+    }
+    node = head;
+    while (ln1 != NULL && ln2 != NULL) {
+        if (ln1->val < ln2->val) {
+            node->next = ln1;
+            ln1 = ln1->next;
+        } else {
+            node->next = ln2;
+            ln2 = ln2->next;
+        }
+        node = node->next;
+    }
+    if (ln1 != NULL) {
+        node->next = ln1;
+    } else {
+        node->next = ln2;
+    }
+    return head;
+}
+
+struct ListNode *mergeKLists(struct ListNode** lists, int lists_size) {
+    if (lists_size == 0) {
+        return NULL;
+    }
+    struct ListNode **ml, *head;
+    int i, is_odd, sz = lists_size;
+    ml = malloc(sizeof(struct ListNode*) * lists_size);
+    for (i = 0; i < lists_size; i++) {
+        ml[i] = lists[i];
+    }
+    while(sz > 1) {
```

Comment:
I think you can use more simple approach:

the function `mergeTwo(target, source)` - merges target and source into the target in the sorted order.
the function `mergeKLists` :
 - creates a result head.
 - uses `for (lists_size)`: 
 -   each loop uses mergeTwo(head, curr_list).

I believe this case you will have the Space complexity: O(1), Runtime: (Sum(listsizes)).

Reply:
Yes, it will use O(1) space complexity, but will use runtime complexity of O(k^2 * n), where k is the numbers of lists, and n is the average list size - such run complexity is lower than the current implementation, which is O(k * n * log(k)).